### PR TITLE
Add scalar tensor

### PR DESCRIPTION
**Pull request by WashimNeupane**
## Description
Added a aten.scalar_tensor op to frontend. 
The requirement to have a persistent session across the SDFG was removed/commented out. Need better fix in another PR. 

## Related Issues

- Closes #

## Type of Change
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactor / internal change (no functional change)
- [ ] Documentation update

## Quality Checklist
*Please ensure the following are completed before requesting review:*

* [x] Runtime-compatible (no externally visible existing method has its parameters changed, was deleted or has its exception behavior changed)
* [x] Compile-time compatible (new arguments & properties added have default values, but users need to recompile against the new version to work)

IR:
* [ ] Introduces new types into the SDFG IR
* [ ]  Changed / added properties on SDFG IR Elements (LibraryNodes, ControlFlowNodes DataFlowNodes etc.)
  *  [ ] New / modified types have serialization / deserialization support
  *  [ ] Deserialization is backward compatible (new serialization can deserialize old format and represent it in the new format if applicable)
  *  [ ] Serialized format is backward compatible (previous deserializer can read new serialized format without incorrectness (just lacking additional guarantees that are optional or did not exist before)
  *  [ ] Serializer has an option to emit the previous format

### Unit Tests
- [x] New unit tests have been added for the changes.
- [x] Existing unit tests pass locally.
- [x] Edge cases and error paths are covered.

### Integration Tests
- [ ] New integration tests have been added (or existing ones updated).
- [ ] Integration tests pass locally.
- [ ] Interactions with other components/services are verified.

### Documentation
- [ ] Public APIs, options, and behavior changes are documented.
- [ ] README / docs site / inline comments updated where applicable.
- [ ] Changelog / release notes updated (if applicable).

### Test Coverage
- [x] Test coverage has not decreased as a result of this change.
- [x] Coverage report reviewed and acceptable.
- [x] Critical / new code paths are covered.

## How Has This Been Tested?
-verified no memory allocation (malloc) is removed for the 0d stack-allocated scalar tensor.

## Additional Context
Add any other context about the PR here (design decisions, trade-offs, follow-ups).

### CODE REVIEW

File: pytorch/tests/ops/test_creation.py [hunk 1]
```diff
@@ -49,3 +49,46 @@ def forward(self, input: torch.Tensor) -> torch.Tensor:
...
+
+    check(ScalarTensorLayoutNet(), target=target)
+
+
+def test_scalar_tensor_add(target: str) -> None:
+    class ScalarTensorAddNet(nn.Module):
+        def forward(self, input: torch.Tensor) -> torch.Tensor:
+            return input + torch.scalar_tensor(3.141592)
+
+    check(ScalarTensorAddNet(), torch.ones(2, 3), target=target)
```
Comment (Atrisan) [inline pytorch/tests/ops/test_creation.py:94]: What about tests for pinned_memory and memory_format?
Are they not supported?

File: pytorch/docc/pytorch/graph_parser/utils.py [hunk 2]
```diff
@@ -899,13 +899,20 @@ def allocate_memory(
                 node,
                 "Expected ContainerInfo but got: " + str(type(info)),
             )
+
+        # Skip allocation for Scalar Tensors
+        if isinstance(info.sdfg_type(), Scalar):
```
Comment (Atrisan) [inline pytorch/docc/pytorch/graph_parser/utils.py]: You are checking for scalars, which should not be allocated.
A Scalar tensor is still of type Tensor, which is why it needs additional coverage.

If a scalar is passed to this function, something is not filtered correctly in the parser.
Thus, if you add this check it should rather throw an error than hide a noop.

File: pytorch/docc/pytorch/graph_parser/creation.py [hunk 3]
```diff
@@ -117,15 +117,91 @@ def parse(
                 node, fill_value, result_tensor.element_type
             )
 
+        container_type = container_info[result_container].sdfg_type()
+        if isinstance(container_type, Pointer) and len(result_tensor.shape) == 0:
```
Comment (Atrisan) [inline pytorch/docc/pytorch/graph_parser/creation.py]: Why do you check for pointers to create a new tensor?
Either this is something to handle in the utils or its somewhat redundant.
It looks fishy

File: pytorch/docc/pytorch/graph_parser/creation.py [hunk 4]
```diff
@@ -117,15 +117,91 @@ def parse(
...
+        container_info: ContainerInfos,
+    ) -> None:
+        if len(node.args) != 1:
+            raise GraphParserError(
+                self,
+                node,
+                f"Expected exactly 1 argument (scalar value), got {len(node.args)}",
+            )
+        if not set(node.kwargs.keys()).issubset(
+            {"dtype", "layout", "device", "pin_memory", "memory_format"}
```
Comment (Atrisan) [inline pytorch/docc/pytorch/graph_parser/creation.py]: kwarg tests are never type checked like the FullParser

**Review by Atrisan — CHANGES_REQUESTED** [review]

File: pytorch/tests/ops/test_creation.py (hunk 1 above)
Comment (WashimNeupane) [reply pytorch/tests/ops/test_creation.py:94]: i was following other tests patterns i.e. fullparser.. they are supported.

File: pytorch/docc/pytorch/graph_parser/utils.py (hunk 2 above)
Comment (WashimNeupane) [reply pytorch/docc/pytorch/graph_parser/utils.py]: When we use a 0-dimensional tensor internally in PyTorch module (like input + torch.scalar_tensor(...) in test_scalar_tensor_add), the docc graph parser realizes it doesn't need to return it, so it maps it to a primitive Scalar container. This guard was to avoid that. However, new changes to the code will not map it to scalar container within this function, so this wont be necessary.

File: pytorch/docc/pytorch/graph_parser/creation.py (hunk 3 above)
Comment (WashimNeupane) [reply pytorch/docc/pytorch/graph_parser/creation.py]: we were initialising a pointer for scalar tensor. this causes weird bugs. this will be resolved by simply always creating a Scalar container (no pointers), which gets changed to a pointer bufffer when required.

File: pytorch/docc/pytorch/graph_parser/creation.py (hunk 4 above)
Comment (WashimNeupane) [reply pytorch/docc/pytorch/graph_parser/creation.py]: added to new commit.

**Review by Atrisan — DISMISSED** [review]

**Review by Atrisan — APPROVED** [review]
